README.md
# Hearty McWeatherface

Do you have a Fitbit Sense or Versa 3? Do You like seeing a crapload of stats in one glance? If you answered 'yes' to both of those questions, and you have pretty ok eyesight, then you might be interested in this watch face.

## This Repository

This repository contains everything needed to build the watch face using Fitbit Studio. The assets used to create the assets are in the assets directory, as one would expect. The studio directory contains a dump of the Fitbit Studio project, because I like to be able to see what horrid bit of code broke what, when, and where.

I also track changes to the watch face as published here in the CHANGELOG, and at some point in the future, I'll probably add a help site for GitHub Pages to point to from the docs. There's a lot on the face that could use more description, but the Fitbit Gallery doesn't allow for very much.


CHANGELOG.md
## 1.2.2

Submitted for approval

* __BUG__: Fixed imperial rendering of elevation gained.

## 1.2.1

Unpublished

* __BUG__: Fixed tracking of min/max long to be min/max of mid min/max when it gets cycled through rather than just the min/max of the mid avg.

## 1.2.0

Unpublished

* Added percentage chance of precipitation to main view weather indicators as a blue bar to the left of the icon that gets taller as the chance goes up.
* Added min/max temperature indicators to the main view weather indicators as red and light blue bars to the right of the icon. Both peak at 10 degrees from the listed temperature at their full height. The light blue bar indicates how much colder it will be for the period, and the red bar indicates how much warmer it will be for the period.
* Added an active zone minutes breakdown on the main view as a bar divided into yellow, orange, and red representing each zone and its percentage of the total.
* Added an active zone minutes breakdown to the stats view inside the total wheel and with a list of the minutes for each to the right.
* Added support for starting, pausing, and stopping exercise:
  * The stats view now has a button at the bottom right to open the exercise view.
  * While ongoing, exercise stats replace the next two days weather section in the main view, and double tapping it switches into the exercise view.
* Exercise view:
  * When not currently exercising;
    * displays an icon representing an exercise and left and right arrow buttons to change the target exercise.
    * The buttons at the bottom cancel (left) to go back to the stats view or play (right) to start the exercise.
  * When an exercise is active:
    * displays the exercise icon at the top
    * to the left of the icon are two weather forecasts for the next six hours in three hour increments starting from the left
    * to the right of the icon is a button that will mark a lap
    * in the band below the icon are the elapsed time of the exercise, the current time, and the battery remaining
    * in the cluster to the left, below the band are elevation gained in m or ft and steps taken
    * in the cluster to the right, below the band are the distance travelled and calories burned
    * the top dial indicates the current pace for a mile with a longer bar indicating a slower time
      * inside the top dial are the current pace and the average pace, on top and bottom, respectively
    * the left dial indicates the current speed
      * the number in the center is current speed
      * the number on the left is average speed
      * the number on the right is max speed
    * the right dial indicates the current heart rate
      * the number in the center is current heart rate
      * the number on the left is average hear rate
      * the number on the right is max heart rate
* Added min/max tracking for mid and long heart rate, indicated by a vertical line over the corresponding point. The mid min/max is gathered from all points when the short rate graph cycles. The long min/max is gathered from all of the min/max points of the mid min/max points when the mid rate graph cycles.
* Added min/max listings to the heart rate view with the max at the top left of the graph and the min at the bottom center.
* Added two timers to the time view, which will trigger a vibration when they elapse and switch to the time view while indicating which timer elapsed.
  * The timer set view is opened with a tap on the timer time slot.
  * Tapping the +/- buttons will add or subtract from the time in their column. Tapping the time will add one to the part that is tapped.
* Switched to the System-Bold font for most text so that the monospace hack can be used.
* Added current stopwatch and timer indicators that appear as triangles on the main view time ':', with the top indicating a running stopwatch and the bottom indicating a running timer.

## 1.1.0

2021-06-29

* Switched weather condition from text to an icon.
* Switched stat indicators to icons.
* Added feels like temperature and percentage chance of precipitation to weather view.
* Made a few consistency adjustments to backgrounds.

## 1.0.0

2021-06-28

Initial port of my Ionic watch face that worked in much the same way over to SDK 5 and the Versa 3 era devices.

Features include:

### Main view

* Big 'ole time right in the middle.
* Not so big date right under it.
* Day of week indicator at the bottom of the main view.
* Heart rate graph on the bottom half main view, split into three sections that represent the last minute and forty seconds, 30 minutes, and 10 hours starting from the right.
* Weather city at the top of the main view.
* Current weather in 3 hour increments in the top left pane. The top left weather is now and, moving clockwise, the other two are three hours out. Weather is gathered from the lovely [OpenWeatherMap](https://openweathermap.org/).
* Next two days of weather in the top center pane.
* Current stats in the top right pane, with the progress bar behind indicating completion of the corresponding goal. From top to bottom:
  * Battery percentage
  * Elevation gained in floors
  * Step count
  * Distance moved
  * Calories burned
  * Total active zone minutes
* Each area of the view can be double tapped to get more details. The areas:
  * Time
  * Stats
  * Heart rate long (bottom left)
  * Heart rate mid (bottom center)
  * Heart rate short (bottom right)
  * Weather

### Weather view

* This view shows detailed forecast for the next three days.
* The city (specified or found by location) is listed at the top.
* To the right of the city is the timestamp that the data was retrieved.
* The fields in the table taking up most of the view are (from left to right):
  * day and hour for the entry
  * condition
  * temperature
  * humidity
  * wind direction and speed
  * percentage of cloud cover
* Double tapping anywhere goes back to the main view.

### Time view


studio/app/demo.js
export const ekg = {
  minLongs: [ 88,  97,  85,  66,  54,  70,  72,  65,  61,  61,  63,  50,  54,  54,  55,  59,  52,  70,  74,  81],
  longs:    [ 97,  99,  88,  86,  71,  71,  72,  68,  69,  69,  65,  50,  56,  57,  56,  59,  63,  75,  80,  85],
  maxLongs: [102, 133, 100,  88,  78,  79,  92,  68,  69,  79,  67,  55,  58,  59,  59,  59,  78,  79, 169, 155],

  minMids:  [ 88,  88,  85,  80,  79,  84,  88,  90,  98,  98, 102, 110, 104, 104, 110, 110, 127, 117, 101,  96],
  mids:     [ 90,  88,  87,  88,  87,  87,  90,  96, 100, 110, 111, 110, 105, 109, 114, 120, 132, 120, 122, 110],
  maxMids:  [100,  99,  99,  95,  93,  94, 110, 120, 110, 115, 114, 130, 135, 149, 144, 130, 172, 121, 125, 111],

  shorts:   [110, 115, 112, 111, 110, 109, 110,  99,  98,  95,  93,  92,  93,  90,  85,  80,  82,  80,  79,  75],
};

export const stats = {
  steps: 6942,
  distance: 4545,
  calories: 1420,
  elevationGain: 8,
  activeZoneMinutes: {
    fatBurn: 10, cardio: 7, peak: 4, total: 21,
  },
};

export const goals = {
  steps: 10100,
  distance: 5000,
  calories: 3000,
  elevationGain: 10,
  activeZoneMinutes: {
    fatBurn: 20, cardio: 5, peak: 0, total: 20,
  },
};

export const exercise = {
  state: 'stopped',
  type: 'run',
  stats: {
    steps: 1621,
    elevationGain: 37,
    speed: {
      current: 2.5,
      average: 2.0,
      max: 3.2,
    },
    pace: {
      current: 563,
      average: 602,
      max: 671,
    },
    heartRate: {
      current: 122,
      average: 108,
      max: 177,
    },
    calories: 308,
    distance: 1500,
    activeTime: 2978141.
  }
};

export const water = {
  amount: 730,
  goal: 2290
};

export const weather = {
  city: 'Demo City',
  when: Date.now() - 3600000,
  list: [
    { temp: 292.61, feel: 292.75, precip: 0.05, desc: 'Clear', min: 290.27, max: 292.61, hum: 82, when: 1625475600, clouds: 0, wind: { speed: 1.06, deg: 89, gust: 1.11 } },
    { temp: 293.8, feel: 293.77, precip: 0.23, desc: 'Clouds', min: 291.8, max: 296.31, hum: 71, when: 1625486400, clouds: 0, wind: { speed: 1.08, deg: 93, gust: 1.77 } },
    { temp: 302.69, feel: 302.13, precip: 0.88, desc: 'Rain', min: 302.69, max: 302.69, hum: 38, when: 1625497200, clouds: 1, wind: { speed: 2.02, deg: 95, gust: 3.26 } },
    { temp: 302.83, feel: 304.51, precip: 0.42, desc: 'Drizzle', min: 302.83, max: 302.83, hum: 55, when: 1625583600, clouds: 96, wind: { speed: 1.16, deg: 61, gust: 1.21 } },
    { temp: 302.61, feel: 304.16, precip: 0, desc: 'Thunderstorm', min: 302.61, max: 302.61, hum: 55, when: 1625605200, clouds: 85, wind: { speed: 2.48, deg: 162, gust: 4.72 } },
    { temp: 295.43, feel: 296.01, precip: 0.45, desc: 'Fog', min: 295.43, max: 295.43, hum: 88, when: 1625670000, clouds: 100, wind: { speed: 3.64, deg: 69, gust: 7.5 } },
    { temp: 294.54, feel: 295.24, precip: 0.3, desc: 'Snow', min: 294.54, max: 294.54, hum: 96, when: 1625691600, clouds: 100, wind: { speed: 3.46, deg: 38, gust: 8.4 } },
    { temp: 301, feel: 302.82, precip: 0, desc: 'Smoke', min: 301, max: 301, hum: 64, when: 1625756400, clouds: 100, wind: { speed: 1.09, deg: 284, gust: 1.56 } },
    { temp: 303.23, feel: 305.36, precip: 0.04, desc: 'Tornado', min: 303.23, max: 303.23, hum: 56, when: 1625778000, clouds: 100, wind: { speed: 3.47, deg: 261, gust: 6.8 } },
  ],
};

export const demo = {
  ekg, exercise, goals, stats, water, weather
};

studio/common/utils.js
export function zeroPad(i) {
  if (i < 10) {
    i = "0" + i;
  }
  return i;
}

export function avg(arr) {
  return Math.round(arr.reduce((a, c) => a + c, 0) / arr.length);
}

export function keys(obj) {
  if (!obj) return [];
  const keys = [];
  for (const k in obj) keys.push(k);
  return keys;
}

export function fill(arr, el, from, to) {
  for (let i = from; i < to; i++) arr[i] = el;
  return arr;
}

export function find(arr, fn) {
  for (let i = 0; i < arr.length; i++) {
    if (fn(arr[i])) return arr[i];
  }
}

export function kelvinToC(temp) {
  return Math.round((+temp * 10) - 2731.5) / 10;
}

export function kelvinToF(temp) {
  return (Math.round(1.8 * ((+temp * 10) - 2731.5)) / 10) + 32
}

export function mToF(m) {
  if (!m) return 0;
  return m * 3.28084;
}

export function mToMi(m) {
  if (!m) return 0;
  return Math.round((m / 1609.344) * 100) / 100;
}

export function mToKm(m) {
  if (!m) return 0;
  return Math.floor(m / 1000);
}

export function mToMiLong(m) {
  return m / 1609.344;
}

export function msToKph(ms) {
  if (!ms) return 0;
  return ms * 3.6;
}

export function msToMph(ms) {
  if (!ms) return 0;
  return mToMiLong(ms * 3600);
}

export function skmToMs(skm) {
  if (!skm) return 0;
  return 1 / (skm / 1000);
}

export function skmToSmi(skm) {
  if (!skm) return 0;
  return 1 / mToMiLong(skmToMs(skm));
}

export function mlToOz(ml) {
  return Math.round(ml / 29.57);
}

export function ozToMl(oz) {
  return Math.round(oz * 29.57);
}

export function formatMilliseconds(ms, padh) {
  if (!ms) return padh ? '0:00:00' : '0:00';
  
  ms = Math.round(ms);
  
  const h = Math.floor(ms / 3600000);
  ms -= h * 3600000;
  const m = Math.floor(ms / 60000);
  ms -= m * 60000;
  const s = Math.floor(ms / 1000);
  
  return `${h ? `${h}:` : (padh ? '0:' : '')}${m ? (h || padh ? zeroPad(m) : m) : (padh ? '00' : '0')}:${zeroPad(s)}`;
}

export function formatSeconds(s, padh) {
  if (!s) return padh ? '0:00:00' : '0:00';
  return formatMilliseconds(s * 1000, padh);
}

const monodigits = [0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19].map(c => String.fromCharCode(c));
const digits = /[0-9]/g;
export function monoDigits(str) {
  return `${str}`.replace(digits, d => monodigits[+d]);
}

export function multiTap(count) {
  return function(el, fn) {
    let clicks = 0;
    let pos;
    let timeout;
    el.addEventListener('mouseup', ev => {
      if (clicks || count === 1) {
        clicks++;
        if (timeout) clearTimeout(timeout);
        timeout = 0;
        if (count === 1 || Math.abs(Math.abs(pos.x - ev.screenX) - Math.abs(pos.y - ev.screenY)) < 40) {
          if (clicks === count) {
            fn();
            clicks = 0;
          } else {
            timeout = setTimeout(() => {
              clicks = 0;
            }, 750);
          }
        }
      } else {
        pos = { x: ev.screenX, y: ev.screenY };
        clicks++;
        timeout = setTimeout(() => {
          clicks = 0
        }, 750);
      }
    });
  };
}

export const singleTap = multiTap(1);
export const doubleTap = multiTap(2);
export const tripleTap = multiTap(3);


studio/companion/index.js
import { me } from "companion";
import * as messaging from "messaging";
import { settingsStorage } from "settings";
import { geolocation } from "geolocation";
import { localStorage } from "local-storage";
import shh from '../secrets.js';

const settings = ['time', 'temp', 'hourColor', 'minuteColor', 'heartrateColor', 'graphColor', 'fitbitAuth'];
const queue = [];

const fitbitOAuth = 'https://api.fitbit.com/oauth2/token';

console.log('starting companion');

function flushQueue() {
  if (messaging.peerSocket.readyState === messaging.peerSocket.OPEN) {
    const msgs = queue.splice(0, queue.length);
    console.log('companion flushing queue');
    msgs.forEach(m => {
      console.log(`companion sending ${JSON.stringify(m)}`);
      messaging.peerSocket.send(m);
    });
  } else {
    console.log(`socket not ready for queue flush`);
  }
}

function sendMessage(type, msg) {
  queue.push({ type, value: msg });
  flushQueue();
}

me.monitorSignificantLocationChanges = true;

me.onsignificantlocationchange = evt => {
  if (settingsStorage.getItem('forceZip') !== 'true') {
    const last = localStorage.getItem('lastWeather') || 0;
    // back up the last time 2.5 hours
    localStorage.setItem('lastWeather', last - 2700000);
    checkWeather();
  }
}

settingsStorage.onchange = evt => {
  console.log('settings changed');
  const weather = ['forceZip', 'altZip', 'weatherKey'].includes(evt.key);

  if (weather && localStorage.getItem('forceZip') != settingsStorage.getItem('forceZip') || localStorage.getItem('altZip') != settingsStorage.getItem('altZip') || localStorage.getItem('weatherKey') != settingsStorage.getItem('weatherKey')) {
    localStorage.setItem('lastWeather', 0);
  }
  localStorage.setItem('forceZip', settingsStorage.getItem('forceZip'));
  localStorage.setItem('altZip', settingsStorage.getItem('altZip'));
  localStorage.setItem('weatherKey', settingsStorage.getItem('weatherKey'));
  
  if (weather) checkWeather();
  sendSettings();
}

messaging.peerSocket.onopen = () => {
  console.log('companion peer open');
  flushQueue();
}

messaging.peerSocket.onerror = err => {
  // Handle any errors
  console.log("Connection error: " + err.code + " - " + err.message);
}

messaging.peerSocket.onclose = () => {
  console.log('companion peer closed');
}

// Listen for the onmessage event
messaging.peerSocket.onmessage = evt => {
  if (!evt.data || !evt.data.type) return;
  if (evt.data.type === 'weather') {
    if (evt.data.force) localStorage.setItem('lastWeather', 0);
    checkWeather(evt.data.value);
  } else if (evt.data.type === 'waterlog') {
    sendWater(evt.data.value);
  } else if (evt.data.type === 'water') {
    getWater();
  } else if (evt.data.type === 'sleep') {
    getSleep();
  }
}

const privateSettings = ['fitbitAuth', 'fitbitOauth'];
const sendSettings = debounce(function sendSettings() {
  const obj = {};
  settings.forEach(k => {
    console.log(`setting ${k}: ${settingsStorage.getItem(k)}`);
    if (~privateSettings.indexOf(k)) return;
    try {
      obj[k] = JSON.parse(settingsStorage.getItem(k));
    } catch (e) {
      obj[k] = settingsStorage.getItem(k);
    }
  });

  try {
    const auth = JSON.parse(settingsStorage.getItem('fitbitAuth'));
    if (auth) {
      obj.canWater = true;
      obj.canSleep = true;
    }
  } catch (e) {
    console.log(`[auth check err] ${e}`);
  }
  
  queue.push({ type: 'settings', value: obj });
  flushQueue();
}, 500);

var ENDPOINT = "https://api.openweathermap.org/data/2.5/forecast";

function zeroPad(n) {
  if (n < 10) return `0${n}`;
  else return `${n}`;
}

function processWeather(response) {
  if (response.status >= 200 && response.status <= 299) {
    response.json().then(function(data) {
      console.log(`got weather for ${JSON.stringify(data.city)}`);
      const weather = { info: {} };
      weather.info.city = data.city.name;
      weather.info.when = Date.now();
      weather.info.loc = data.coords;
      weather.list = data.list.map(w => ({
        temp: w.main.temp,
        feel: w.main.feels_like,
        precip: w.pop,
        desc: w.weather[0].main,
        id: w.weather[0].id,
        min: w.main.temp_min,
        max: w.main.temp_max,
        hum: w.main.humidity,
        when: w.dt,
        clouds: w.clouds.all,
        wind: w.wind
      }));
      const list = weather.list;
      console.log(JSON.stringify(list[0]))
      
      let morning = +(new Date(list[0].when * 1000)).toString().substr(16, 2);
      while (morning + 3 <= 12) morning += 3;
      while (morning > 12) morning -= 3;
      let evening = morning + 6;
      
      const tmp = [(new Date((list[0].when * 1000) + 86400000)).toString().substr(0, 16), (new Date((list[0].when * 1000) + 172800000)).toString().substr(0, 16), (new Date((list[0].when * 1000) + 259200000)).toString().substr(0, 16)];
      const dts = [];
      for (let i = 0; i < tmp.length; i++) dts.push(`${tmp[i]}${zeroPad(morning)}`, `${tmp[i]}${zeroPad(evening)}`);
      weather.list = list.reduce((a, c, i) => {
        if (i <= 2) a.push(c);
        else if (~dts.indexOf((new Date(c.when * 1000)).toString().substr(0, 18))) a.push(c);
        return a;
      }, []);
      
      localStorage.setItem('lastWeather', weather.info.when);
      localStorage.setItem('weather', JSON.stringify(weather));
      sendWeather();
    });
  }
}

function sendWeather() {
  const weather = JSON.parse(localStorage.getItem('weather'));
  queue.push({ type: 'weather', value: weather.info });
  weather.list.forEach((l, i) => queue.push({ type: 'weather', key: i, value: l }));
  flushQueue();
}

function json(str) {
  let res = {};
  try {
    res = JSON.parse(str);
  } catch (e) {}
  if (!res) res = {};
  return res;
}

// Fetch the weather from OpenWeather
function queryOpenWeather() {
  const key = json(settingsStorage.getItem('weatherKey')).name || shh.openWeatherKey;
  console.log(`getting weather with key ${key}`)
  if (localStorage.getItem('forceZip') === 'true') {
    console.log(`fetching weather for zip ${settingsStorage.getItem('altZip')}`);
    fetch(ENDPOINT + `?zip=${json(settingsStorage.getItem('altZip')).name || '90210'}&appid=${key}`, { mode: 'cors' }).then(processWeather, err => {
      console.error("Error fetching weather: " + err);
    });
  } else {
    geolocation.getCurrentPosition(position => {
      fetch(ENDPOINT + `?lat=${position.coords.latitude}&lon=${position.coords.longitude}&appid=${key}`, { mode: 'cors' })
      .then(processWeather)
      .catch(function (err) {
        console.log("Error fetching weather: " + err);
      });
    }, err => {
      fetch(ENDPOINT + `?zip=${json(settingsStorage.getItem('altZip')).name || '90210'}&appid=${key}`, { mode: 'cors' }).then(processWeather, err => {
        console.error("Error fetching weather: " + err);
      });
    });
  }
}

const checkWeather = debounce(function checkWeather(value) {
  const last = JSON.parse(localStorage.getItem('lastWeather'));
  if (!last || last < (Date.now() - 3600000)) {
    console.log(Date.now() - 3600000, last);
    queryOpenWeather();
  } else {
    if (!value || value != last) sendWeather();
  }
}, 500);

function fitbitKey() {
  return JSON.parse(settingsStorage.getItem('fitbitAuth')).access_token;
}

function today() {
  return new Date().toISOString().substr(0, 10);
}

function sendWater(ml) {
  oauthFetch('fitbitAuth', fitbitOAuth, `https://api.fitbit.com/1/user/-/foods/log/water.json?amount=${ml}&date=${today()}&unit=ml`, () => ({
    method: 'POST',
    headers: {
      Authorization: `Bearer ${fitbitKey()}`,
    },
  }))
  .then(data => {
    console.log(JSON.stringify(data))
    if (data.waterLog && data.waterLog.logId) getWater();
  })
  .catch(err => console.log('[send water api]: ' + err));
};

function getWater() {
  oauthFetch('fitbitAuth', fitbitOAuth, `https://api.fitbit.com/1/user/-/foods/log/date/${today()}.json`, () => ({
    method: "GET",
    headers: {
      Authorization: `Bearer ${fitbitKey()}`,
    }
  }))
  .then(food => {
    oauthFetch('fitbitAuth', fitbitOAuth, `https://api.fitbit.com/1/user/-/foods/log/water/goal.json`, () => ({
      method: "GET",
      headers: {
        Authorization: `Bearer ${fitbitKey()}`,
      }
    }))
    .then(water => {
      sendMessage('water', { amount: food.summary.water, goal: water.goal.goal });
    });
  })
  .catch(err => console.log('[get water api]: ' + err));
}

function getSleep() {
  oauthFetch('fitbitAuth', fitbitOAuth, `https://api.fitbit.com/1.2/user/-/sleep/date/${today()}.json`, () => ({
    method: "GET",
    headers: {
      Authorization: `Bearer ${fitbitKey()}`,
      'Accept-Language': 'en_CA'
    }
  }))
  .then(res => res.json())
  .then(sleep => {
    oauthFetch('fitbitAuth', fitbitOAuth, `https://api.fitbit.com/1/user/-/sleep/goal.json`, () => ({
      method: "GET",
      headers: {
        Authorization: `Bearer ${fitbitKey()}`,
      }
    }))
    .then(goal => {
      const main = sleep.sleep.find(s => s.isMainSleep) || sleep.sleep[0];
      const sum = main.levels.summary;
      sendMessage('sleep', {
        goal: goal.goal.minDuration,
        score: main.efficiency,
        asleep: main.minutesAsleep,
        awake: main.minutesAwake,
        start: main.startTime,
        deep: { count: sum.deep.count, minutes: sum.deep.minutes, avg: sum.deep.thirtyDayAvgMinutes },
        light: { count: sum.light.count, minutes: sum.light.minutes, avg: sum.light.thirtyDayAvgMinutes },
        rem: { count: sum.rem.count, minutes: sum.rem.minutes, avg: sum.rem.thirtyDayAvgMinutes },
        wake: { count: sum.wake.count, minutes: sum.wake.minutes, avg: sum.wake.thirtyDayAvgMinutes },
      });
    });
  })
  .catch(err => console.log('[get sleep api]: ' + err));
}

sendSettings();

const refreshAuth = 'Basic ' + btoa(shh.fitbitClientId + ':' + shh.fitbitClientSecret);
console.log(refreshAuth);

async function oauthFetch(auth, oauthurl, url, options, retry = 0) {
  if (!settingsStorage.getItem(auth)) throw new Error(`No auth for ${auth}`);

  let o = options;
  if (typeof options === 'function') o = options();
  const res = await fetch(url, o);
  const data = await res.json();
  if (('success' in data && !data.success) || 'errors' in data) {
    console.log(`[oauth fetch err]\n\t${JSON.stringify(data)}\n\t${url} try ${retry + 1}\n\t${JSON.stringify(o)}\n\t${settingsStorage.getItem(auth)}`);
    if (retry) throw new Error('oauth too many fails');
    await reauth(auth, oauthurl);
    return oauthFetch(auth, url, options, retry + 1);
  }
  return data;
}

async function reauth(auth, oauthurl) {
  const obj = JSON.parse(settingsStorage.getItem(auth));
  console.log(`original: ${JSON.stringify(obj)}`);
  const reup = await fetch(
    `${oauthurl}?grant_type=refresh_token&refresh_token=${obj.refresh_token}`,
    {
      method: 'POST',
      headers: {
        Authorization: refreshAuth,
        'Content-Type': 'application/x-www-form-urlencoded',
      }
    }
  );
  const reupVal = await reup.json();
  console.log(`reup: ${JSON.stringify(reupVal)}`);
  if (JSON.stringify(reupVal).match(/refresh token invalid/i)) {
    console.log(`oauth reup failed due to invalid token, clearing oauth token`);
    settingsStorage.removeItem(auth);
    sendSettings();
  } else {
    if (reupVal.refresh_token) obj.refresh_token = reupVal.refresh_token;
    if (reupVal.access_token) obj.access_token = reupVal.access_token;
    settingsStorage.setItem(auth, JSON.stringify(obj));
  }
}

function debounce(fn, time) {
  let tm;
  return function(...args) {
    if (tm) clearTimeout(tm);
    setTimeout(() => {
      tm = 0;
      fn.apply(this, args);
    }, time);
  }
}

studio/app/elements.js
import document from 'document';

export function eid(id) {
  return document.getElementById(id);
}

export function eclass(name) {
  return document.getElementsByClassName(name);
}

export const ui = {
  page: {
    aod: eid('aod'),
    main: eid('main'),
    heartl: eid('heart-long'),
    weather: eid('weather'),
    stats: eid('stats'),
    time: eid('time'),
    exercise: eid('exercise'),
    water: eid('water'),
    timer: eid('timer')
  },
  aod: {
    wrapper: eid('aod-wrapper'),
    hour1: eclass('aod-hour1'),
    hour2: eclass('aod-hour2'),
    minute1: eclass('aod-minute1'),
    minute2: eclass('aod-minute2'),
    date: eid('aod-date'),
    lines: eclass('aod-ekg-line'),
  },
  main: {
    hours: eid("hours"),
    minutes: eid("minutes"),
    hoursShadow: eid("hours-shadow"),
    minutesShadow: eid("minutes-shadow"),
    date: eid("date"),
    dateShadow: eid('date-shadow'),
    battery: eid("battery"),
    steps: eid("steps"),
    stepsBar: eid("steps-bar"),
    floors: eid("floors"),
    floorsBar: eid("floors-bar"),
    calories: eid("calories"),
    caloriesBar: eid("calories-bar"),
    distance: eid("distance"),
    distanceBar: eid("distance-bar"),
    activeMinutes: eid("active-minutes"),
    activeMinutesBar: eid("active-minutes-bar"),
    activeFatburnBar: eid('active-fatburn-bar'),
    activeCardioBar: eid('active-cardio-bar'),
    activePeakBar: eid('active-peak-bar'),
    minRate: eid("min-rate"),
    minRateShadow: eid("min-rate-shadow"),
    maxRate: eid("max-rate"),
    maxRateShadow: eid("max-rate-shadow"),
    city: eid("city"),
    dayText: eclass("day-text"),
    dayBox: eclass("day-box"),
    heart: {
      lines: eclass('ekg-line-main'),
      ranges: eclass('ekg-range')
    },
    weather: eid('weather'),
    temps: eclass('main-temp'),
    descs: eclass('main-desc'),
    imgs: eclass('weather-img'),
    precips: eclass('weather-precip-bar'),
    lows: eclass('weather-low-bar'),
    highs: eclass('weather-high-bar'),
    stats: eid('stats'),
    weather2block: eid('main-weather-2'),
    weather1: eid('weather-plus1'),
    weather2: eid('weather-plus2'),
    exercise: {
      main: eid('main-exercise'),
      which: eid('main-exercise-icon'),
      steps: eid('main-exercise-steps'),
      calories: eid('main-exercise-calories'),
      distance: eid('main-exercise-distance'),
      floors: eid('main-exercise-floors'),
      time: eid('main-exercise-time'),
      speed: eid('main-exercise-speed')
    },
    activeTm: eid('active-tm'),
    activeSw: eid('active-sw')
  },
  heart: {
    long: {
      min: eid('ekg-l-min'),
      max: eid('ekg-l-max'),
      lines: eclass('ekg-line-l'),
      texts: eclass('ekg-text-l'),
      shadows: eclass('ekg-text-shadow'),
      ranges: eclass('ekg-range-l'),
      time: eid('heart-long-time'),
      battery: eid('heart-long-battery'),
      l20: eid('heart-20'),
      l40: eid('heart-40'),
      l60: eid('heart-60'),
      l80: eid('heart-80')
    }
  },
  weather: {
    time: eid('weather-time'),
    battery: eid('weather-battery'),
    where: eid('weather-where'),
    when: eid('weather-when'),
    whens: eclass('weather-when'),
    feels: eclass('weather-feel'),
    precips: eclass('weather-precip'),
    temps: eclass('weather-temp'),
    clouds: eclass('weather-cloud'),
    hums: eclass('weather-hum'),
    descs: eclass('weather-desc'),
    imgs: eclass('weather-dimg'),
    winds: eclass('weather-wind')
  },
  stats: {
    time: eid('stats-time'),
    battery: eid('stats-battery'),
    pct: {
      step: eid('step-pct'),
      floor: eid('floor-pct'),
      dist: eid('dist-pct'),
      active: eid('active-pct'),
      cal: eid('cal-pct'),
      fatburn: eid('fatburn-pct'),
      cardio: eid('cardio-pct'),
      peak: eid('peak-pct')
    },
    amt: {
      step: eid('step-amt'),
      floor: eid('floor-amt'),
      dist: eid('dist-amt'),
      active: eid('active-amt'),
      cal: eid('cal-amt'),
      fatburn: eid('fatburn-amt'),
      cardio: eid('cardio-amt'),
      peak: eid('peak-amt')
    },
    goal: {
      step: eid('step-goal'),
      floor: eid('floor-goal'),
      dist: eid('dist-goal'),
      active: eid('active-goal'),
      cal: eid('cal-goal'),
      fatburn: eid('fatburn-goal'),
      cardio: eid('cardio-goal'),
      peak: eid('peak-goal')
    },
    exercise: eid('stats-exercise'),
    water: {
      wrapper: eid('stat-water-wrapper'),
      button: eid('stats-water'),
      back: eid('stat-water-back'),
      bar: eid('stat-water'),
      text: eid('stat-water-text'),
      goal: eid('stat-water-goal'),
    },
  },
  time: {
    battery: eid('time-battery'),
    time: eid('time-time'),
    date: eid('time-date'),
    sw1: eid('time-sw1'),
    sw2: eid('time-sw2'),
    tm1: eid('time-tm1'),
    tm2: eid('time-tm2')
  },
  exercise: {
    bg: eid('exercise-bg'),
    activeEl: eid('exercise-active'),
    pickEl: eid('exercise-pick'),
    battery: eid('exercise-battery'),
    time: eid('exercise-clock'),
    active: {
      which: eid('exercise-active-icon'),
      stop: eid('exercise-stop'),
      btn: eid('exercise-btn'),
      btnImg: eid('exercise-btn-img'),
      lap: eid('exercise-lap'),
      laps: eid('exercise-laps'),
      
      time: eid('exercise-time'),
      floors: eid('exercise-floors'),
      calories: eid('exercise-calories'),
      speed: eid('exercise-speed'),
      distance: eid('exercise-distance'),
      steps: eid('exercise-steps'),
      
      speedo: eid('exercise-speedo'),
      speedAvg: eid('exercise-speed-avg'),
      speedMax: eid('exercise-speed-max'),
      
      heart: eid('exercise-heart'),
      hearto: eid('exercise-hearto'),
      heartAvg: eid('exercise-heart-avg'),
      heartMax: eid('exercise-heart-max'),
      zone: {
        normal: eid('exercise-heart-normal'),
        normalLine: eid('exercise-heart-normal-line'),
        fatburn: eid('exercise-heart-fatburn'),
        fatburnLine: eid('exercise-heart-fatburn-line'),
        cardio: eid('exercise-heart-cardio'),
        cardioLine: eid('exercise-heart-cardio-line'),
        peak: eid('exercise-heart-peak'),
        peakLine: eid('exercise-heart-peak-line'),
      },
      
      pace: eid('exercise-pace'),
      paceo: eid('exercise-paceo'),
      paceAvg: eid('exercise-pace-avg'),
      
      temps: eclass('exercise-temp'),
      descs: eclass('exercise-desc'),
      imgs: eclass('exercise-weather-img'),
      precips: eclass('exercise-precip-bar'),
      lows: eclass('exercise-low-bar'),
      highs: eclass('exercise-high-bar'),
    },
    pick: {
      which: eid('exercise-pick-icon'),
      left: eid('exercise-left'),
      right: eid('exercise-right'),
      ok: eid('exercise-ok'),
      cancel: eid('exercise-cancel')
    }
  },
  water: {
    oz8: eid('water-pick-8'),
    oz12: eid('water-pick-12'),
    oz16: eid('water-pick-16'),
    oz20: eid('water-pick-20'),
    oz32: eid('water-pick-32'),
    cancel: eid('water-cancel'),
  },
  timer: {
    time: eid('timer-time'),
    hplus1: eid('timer-h-p1'),
    hplus5: eid('timer-h-p5'),
    hminus1: eid('timer-h-m1'),
    hminus5: eid('timer-h-m5'),
    mplus1: eid('timer-m-p1'),
    mplus5: eid('timer-m-p5'),
    mminus1: eid('timer-m-m1'),
    mminus5: eid('timer-m-m5'),
    splus1: eid('timer-s-p1'),
    splus5: eid('timer-s-p5'),
    sminus1: eid('timer-s-m1'),
    sminus5: eid('timer-s-m5'),
    ok: eid('timer-ok'),
    cancel: eid('timer-cancel')
  }
};
